db.go

```diff
@@ -0,0 +1,236 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	_ "github.com/mattn/go-sqlite3"
+
+	"github.com/uptrace/bun"
+	"github.com/uptrace/bun/extra/bundebug"
+	"github.com/uptrace/bun/dialect/sqlitedialect"
+
+)
+
+// a DBLike is either a *bun.DB or a *bun.Tx
+type DBLike interface {
+	NewInsert() *bun.InsertQuery
+	NewSelect() *bun.SelectQuery
+	NewUpdate() *bun.UpdateQuery
+}
+
+// DB is our OO-style database interface
+// Both RootDB and TxDB implement this, but:
+//  - RootDB has a .Begin() method that returns a TxDB
+//  - TxDB has .Commit() and .Rollback() methods
+type DB interface {
+	UpsertBuild(b Build) error
+	GetBuild(buildNum int) (*Build, error)
+	GetBuildsForPull(pull int) ([]Build, error)
+	OldestUnfinishedBuild() (*Build, error)
+	LatestFinishedBefore(before *Build) (*Build, error)
+	GetArchivableBuild() (*Build, error)
+	UnarchivedBuilds() ([]Build, error)
+	ArchiveBuild(buildNum int) error
+	UpsertArtifacts(artifacts []Artifact) error
+	GetArtifacts(buildNum int) ([]Artifact, error)
+}
+
+
+// dbImpl implements our OO-style database interface
+// dbImpl automatically implements DBLike by composition.
+type dbImpl struct {
+	DBLike
+}
+
+// RootDB implements DBLike and dbImpl.
+// Additionally RootDB has a Begin() method.
+type RootDB struct {
+	// bun.DB makes RootDB implement DBLike
+	*bun.DB
+	// dbImpl will just contain another pointer to the *bun.DB
+	dbImpl
+}
+
+// TxDB provides the same OO-style database, plus the usual bun.Tx functions
+// TxDB implements DBLike and dbImple.
+// Additionally it has .Rollback() and .Commit() mechanisms offered by bun.Tx.
+type TxDB struct {
+	*bun.Tx
+	// dbImpl will just contain another pointer to the *bun.Tx
+	dbImpl
+}
```

Comment:
Ok, I tried harder to get the desired behavior.

I like this a lot.  The only real downside is the need for the interface... but we already have that in the master for other reasons.  If you like this too, I'll likely start with something like this in the master.

Reply:
going with @ioga's [strategy](https://github.com/determined-ai/ci-shortcuts/pull/2) instead